File: Swift/ComplexNumber.swift
let e = 2.718281828459045
let pi = 3.141592653589793238

struct Complex: CustomStringConvertible {
  var real: Double
  var imag: Double
  var description: String {
    let im = imag > 0 ? " + \(imag)i" : imag == 0 ? "" : " - \(-imag)i"
    return "\(real)\(im)"
  }
  var size: Double {
    return (real*real + imag*imag).squareRoot()
  }
  var conj: Complex {
    return Complex(real: real, imag: -imag)
  }
  var inv: Complex {
    let s = size*size
    return conj/s
  }

  static func +(this: Complex, other: Complex)-> Complex {
    return Complex(
      real: this.real+other.real,
      imag: this.imag+other.imag
    )
  }

  static func +(this: Complex, other: Double)-> Complex {
    return Complex(
      real: this.real+other,
      imag: this.imag
    )
  }

  static func +(this: Double, other: Complex)-> Complex {
    return other+this
  }

  static func *(this: Complex, other: Double)-> Complex {
    return Complex(
      real: this.real*other,
      imag: this.imag*other
    )
  }

  static func *(this: Double, other: Complex)-> Complex {
    return other*this
  }

  static func *(this: Complex, other: Complex)-> Complex {
    let a = this.real, b = this.imag,
        c = other.real, d = other.imag
    return Complex(
      real: a*c-b*d,
      imag: a*d+b*c
    )
  }

  static func -(this: Complex, other: Complex)-> Complex {
    return this + other*(-1)
  }

  static func -(this: Complex, other: Double)-> Complex {
    return this + other*(-1)
  }

  static func -(this: Double, other: Complex)-> Complex {
    return (-1)*(other - this)
  }

  static func /(this: Complex, other: Double)-> Complex {
    return this*(1/other)
  }

  static func /(this: Double, other: Complex)-> Complex {
    return this*other.inv
  }

  static func /(this: Complex, other: Complex)-> Complex {
    return this*other.inv
  }
}

let 𝖎 = Complex(real: 0, imag: 1)

let a = 3-4*𝖎
print(a)

File: Rust/ComplexNumber.rs
struct Complex {
  real: f32,
  imag: f32
}

impl Complex {
  fn new(real: f32, imag: f32) -> Complex {
    Complex { real, imag }
  }

  fn new_unit(angle: f32) -> Complex {
    Complex::new(angle.cos(), angle.sin())
  }

  fn size(&self) -> f32 {
    f32::powf(f32::powi(self.real, 2) + f32::powi(self.imag, 2), 0.5)
  }

  fn add(&self, other: Complex) -> Complex {
    Complex::new(self.real+other.real, self.imag+other.imag)
  }

  fn mul(&self, other: Complex) -> Complex {
    let a = self.real;
    let b = self.imag;
    let c = other.real;
    let d = other.imag;
    Complex::new(a*c-b*d, b*c+a*d)
  }

  fn addr(&self, other: f32) -> Complex {
    Complex::new(self.real+other, self.imag)
  }

  fn mulr(&self, other: f32) -> Complex {
    Complex::new(self.real*other, self.imag*other)
  }

  fn conj(&self) -> Complex {
    Complex::new(self.real, -self.imag)
  }
  
  fn angle(&self) -> f32 {
    self.imag.atan2(self.real)
  }

  fn pow(&self, n: f32) -> Complex {
    let t = self.angle();
    let r = self.size();
    Complex::new_unit(n*t).mulr(f32::powf(r, n))
  }
}

fn main() {
  let c = Complex::new(1.0, 2.0).mulr(2.0).addr(1.0);
  println!("{}", c.pow(2.0).size());
  assert_eq!(
    c.size()*c.size(),
    c.mul(c.conj()).real
  )
}
